### PR TITLE
Create mount_option_home template and use it in ol8

### PR DESCRIPTION
**Pull request by Xeicker**
#### Description:

- Create `mount_option_home` template, this searches in which partitions are located the interactive users and checks that all those partitions have the expected mount option
- Migrate rules `mount_option_home_noexec` and `mount_option_home_nosuid` to a new template for OL8

#### Rationale:

- Covers better the expected behavior of checking the partitions where are the home directories, to have the expected mount options

### CODE REVIEW

**Comment by vojtapolasek** [comment]
Hello, could you please clarify what  does this new template improve when compared to original mount_option template? I think it would be better to bring improvements directly into the mount_option template than creating individual templates for each partition.

**Comment by vojtapolasek** [comment]
Hello @Xeicker  this looks really interesting. I have discovered a glitch - I created a fake user having a home directory at /var/tmp. The variable with mount points then contained both /var and /var/tmp. Could you fix it? I think it needs a tweak to the regex.

**Comment by Xeicker** [comment]
@vojtapolasek it is not that easy, because if `/var/tmp` wasn't a separate partition, then OVAL should look for the mount point `/var`. So the issue at the end is how to find the longest match, I wasn't able to solve that

**Comment by marcusburghardt** [comment]
I have some concerns with this approach. Checking the latest STIG benchmarks for RHEL8 and OL8, I confirmed they are almost the same, except the RHEL8 requirement includes an extra note:
```
Note: If a separate file system has not been created for the user home directories (user home directories are mounted under "/"), this is automatically a finding as the "noexec" option cannot be used on the "/" system.
```
The case mentioned by @vojtapolasek  could be slightly changed to a more risky scenario. For example, if an interactive user (UID >= 1000) has the home directory set, by any reason, to `/lib, /opt, /usr or even /`, this remediation would be capable to break the system, for example, if binary execution is blocked in a critical partition. I am not sure for now if the system has any safe guard for this, but I would be more conservative.

I think the proper solution for this would be a an explicit recommendation from the benchmark restricting the partitions where the interactive users have a home directory. Something like, `no home directory for interactive users on default mount points different than /home`.

**Comment by Xeicker** [comment]
@marcusburghardt well in that case I think it is reasonable to add a check in OVAL to verify that all interactive users have their home directories into `/home`. What do you think?

**Comment by marcusburghardt** [comment]
> @marcusburghardt well in that case I think it is reasonable to add a check in OVAL to verify that all interactive users have their home directories into `/home`. What do you think?

I think it could be good. However, it looks more feasible to work with an exclusion list instead of an inclusion list. For example, instead of checking if the user home directory is in `/home` partition, I would check if the user home directly **is not** in any critical partition.

File: shared/templates/mount_option_home/oval.template [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+                                  "/usr",
+                                  "/bin",
+                                  "/sbin",
+                                  "/boot",
+                                  "/dev",
+                                  "/proc"] %}}
+<def-group>
+    <definition class="compliance" id="{{{ _RULE_ID }}}" version="1">
+        {{{ oval_metadata("Home mount points should be mounted with mount option " + MOUNTOPTION + ".") }}}
+        <criteria>
```
Comment (marcusburghardt) [inline shared/templates/mount_option_home/oval.template:12]: Please, include explicit `operator` and a `comment`, so it helps reading logs and debugging.

File: shared/templates/mount_option_home/oval.template [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+            object_ref="object_home_partition_all_partitions_{{{ MOUNTOPTION }}}" />
+            <literal_component>).*?:.*</literal_component>
+        </concat>
+    </local_variable>
+
+    <local_variable id="var_home_partitions_mount_points_{{{ MOUNTOPTION }}}" datatype="string" version="1"
+    comment="This variable collects all partition mount points for interactive users">
+        <object_component item_field="subexpression"
+        object_ref="object_home_partitions_mount_points_{{{ MOUNTOPTION }}}" />
+    </local_variable>
```
Comment (marcusburghardt) [inline shared/templates/mount_option_home/oval.template:45]: The idea used here was to collect all partitions and then filter which ones are related to interactive users. I would suggest a different approach of identifying the interactive users first and then checking the respective partitions, as already used in other similar rules, like here:
https://github.com/ComplianceAsCode/content/blob/master/linux_os/guide/system/accounts/accounts-session/accounts_user_interactive_home_directory_exists/oval/shared.xml

This suggested approach is less complex and more flexible since the range of uids reserved for interactive users can change in different distros. Also, considering servers, it seems more likely to have less interactive users than partitions.servers, it seems more likely to have less interactive users than partitions. Therefore, it is expected this suggested approach would improve a little the template performance.

**Review by marcusburghardt — CHANGES_REQUESTED** [review]

File: shared/templates/mount_option_home/oval.template (hunk 2 above)
Comment (Xeicker) [reply shared/templates/mount_option_home/oval.template:45]: I also thought that your suggestion was a better approach. But I just found several ways on how it doesn't work.

My issue, basically was that that I didn't found a way to filter based on "is substring of". That way I could only include those partitions which their mount point is substring of any home directory

Comment (marcusburghardt) [reply shared/templates/mount_option_home/oval.template:47]: I see. So, basically, the intention is:
1 - Collect a list home directories owned by interactive users
2 - Collect a list of partitions mount points
3 - Filter the mount points list to contain only mount points used for home directories of interactive users
4 - Check the mount options for the filtered list of mount points

I have checked the OVAL documentation.
It is possible to use `password_object` to collect the home directories and satisfy the `Step 1`, as already mentioned. Example:
https://github.com/ComplianceAsCode/content/blob/master/linux_os/guide/system/accounts/accounts-session/accounts_user_interactive_home_directory_exists/oval/shared.xml#L12-L29

The `Step 2` seems good as already defined in this PR.

The `Step 3` is the challenge. I believe the OVAL operation `subset of` could solve the problem. However, I have never tried it and some tests would be necessary to confirm. I think it is worth to test since it would permit a simpler solution if it works:
https://oval-community-guidelines.readthedocs.io/en/latest/oval-schema-documentation/oval-common-schema.html#operationenumeration-1

The `Step 4` seems good as already defined in this PR.

File: shared/templates/mount_option_home/oval.template [hunk 3]
```diff
@@ -0,0 +1,77 @@
+{{% set non_allowed_partitions = ["/lib",
```
Comment (marcusburghardt) [inline shared/templates/mount_option_home/oval.template:1]: In a situation where the interactive user home directory is a sub-folder of a "not allowed" mount point, the check wouldn't pass even if the partition have the mount `option` set. Maybe the template should not consider the "not allowed" partitions for the check. The main risk is about the remediation, which shouldn't change critical partitions.

So, I believe the OVAL check could be simplified by removing the `test_home_allowed_partition_{{{ MOUNTOPTION }}}` and keeping this control only in the remediation level. I am open to more opinions.

File: shared/templates/mount_option_home/oval.template [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    comment="Home mount points aren't in unexpected partition"
+    id="test_home_allowed_partition_{{{ MOUNTOPTION }}}" version="1">
+      <ind:object object_ref="object_home_allowed_partition_{{{ MOUNTOPTION }}}" />
+      <ind:state state_ref="state_home_allowed_partition_{{{ MOUNTOPTION }}}" />
+    </ind:variable_test>
+    <ind:variable_object id="object_home_allowed_partition_{{{ MOUNTOPTION }}}" version="1">
+      <ind:var_ref>var_home_partitions_mount_points_{{{ MOUNTOPTION }}}</ind:var_ref>
+    </ind:variable_object>
+    <ind:variable_state comment="unexpected partition"
+    id="state_home_allowed_partition_{{{ MOUNTOPTION }}}" version="1">
```
Comment (marcusburghardt) [inline shared/templates/mount_option_home/oval.template:72]: We should keep the indentation aligned to other templates and the line sizes aligned to the [style guide](https://complianceascode.readthedocs.io/en/latest/manual/developer/04_style_guide.html).

**Comment by marcusburghardt** [comment]
It would also be great to have test scenarios in this template, so we can better test the affected rules. It is possible to use the `mount_option_home_nosuid` rule as reference.

File: shared/templates/mount_option_home/oval.template (hunk 2 above)
Comment (Xeicker) [reply shared/templates/mount_option_home/oval.template:47]: I think it is not possible to use `subset of` as th OVAL documentation says here https://oval-community-guidelines.readthedocs.io/en/latest/oval-schema-documentation/oval-common-schema.html#simpledatatypeenumeration-1

> The string datatype describes standard string data. This datatype conforms to the W3C Recommendation for string data. Expected operations within OVAL for string values are ‘equals’, ‘not equal’, ‘case insensitive equals’, ‘case insensitive not equal’, ‘pattern match’.

I also tried to use it and didn't work, I was able to build ol8 product with this operation applied to a string, so apparently it is syntactically correct, but it doesn't behave as expected

Comment (marcusburghardt) [reply shared/templates/mount_option_home/oval.template:47]: Thanks for sharing this experience with the `subset of`. I sincerely didn't have time to test it and didn't find any implementation of it. It is possible the scanner (openscap) would need some adjustments to align the expected behavior, but I am also not sure about this at the moment.

In this case, your approach of filtering a `textfilecontent54_object` object seems the most reasonable for the moment.

@yuumasato , what do you think? Do you know any OVAL trick worth to try?

File: shared/templates/mount_option_home/bash.template [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+function perform_remediation (){
+
+    {{{ bash_ensure_mount_option_in_fstab( "$1" , MOUNTOPTION , FILESYSTEM , TYPE ) | indent(4) }}}
+
+    {{{ bash_ensure_partition_is_mounted( "$1" ) | indent(4) }}}
+}
+
+readarray -t all_fstab_mount_points  < \
+    <(awk -F':' '{if ($3>=1000 && $3!= 65534) print $6}' /etc/passwd \
+        | xargs df \
```
[the remark this answers is not in the document]
Comment (Xeicker) [reply shared/templates/mount_option_home/bash.template:22]: I found this command on the internet as a way to find which partition a directory is mounted on, here an example output from where I've been testing:
```
$ df /var/home/user1
Filesystem         1K-blocks   Used Available Use% Mounted on
/dev/mapper/ol-var   4184064 588420   3595644  15% /var
```
and int this system indeed `/var` is a valid mount point

File: shared/templates/mount_option_home/oval.template (hunk 3 above)
Comment (Xeicker) [reply shared/templates/mount_option_home/oval.template]: agree

**Comment by Xeicker** [comment]
@marcusburghardt I included tests and made a few updates based on the development of those tests

 * OVAL pass in case no partition with interactive users is found, this could imply, there are no interactive users, or the partition isn't mounted
* Bash first validates if the homedirectory exists, before getting the partition. This handles the situation in which the partition for this directory isn't mounted, or the directory was deleted without updating user info

**Review by marcusburghardt — CHANGES_REQUESTED** [review]
This template is relatively complex due to the dynamic of interactive users. But it already evolved considerably. For now, I have only some concerns about the remediation. There are cases where this chosen approach is not robust enough. I suggest to improve the part where the mount point is identified to be less dependent of user home directory existence and not assume that there will always exist a mount point for the home directories.

File: linux_os/guide/system/permissions/partitions/mount_option_home_nosuid/tests/fstab.fail.sh [hunk 6]
```diff
@@ -1,4 +1,9 @@
 #!/bin/bash
+{{% if "ol" in product %}}
+# platform = multi_platform_example
```
[the remark this answers is not in the document]
Comment (marcusburghardt) [reply linux_os/guide/system/permissions/partitions/mount_option_home_nosuid/tests/fstab.fail.sh:3]: Hum, maybe it was done this way because these tests scenarios are failing. They are failing because the remediation is aborted when the home directory does not exist, which doesn't seem ideal. It would be better to make the remediation more robust instead of skipping test scenario scripts.

[the remark this answers is not in the document]
Comment (Xeicker) [reply linux_os/guide/system/permissions/partitions/mount_option_home_nosuid/tests/fstab.fail.sh:3]: Well I made an equivalent test for the template, and I just thought it would be better to use this new generalized test

File: shared/templates/mount_option_home/bash.template [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+    {{{ bash_ensure_partition_is_mounted( "$1" ) | indent(4) }}}
+}
+
+readarray -t home_directories  < \
+    <(awk -F':' '{if ($3>=1000 && $3!= 65534) print $6}' /etc/passwd )
+
+
+for home_directory in "${home_directories[@]}"
+do
+    if [ -d $home_directory ]; then
```
[the remark this answers is not in the document]
Comment (Xeicker) [reply shared/templates/mount_option_home/bash.template:27]: Well kind of... because in the case that the folder doesn't exists, how would the script know where should it be mounted? Or if for some reason the directory was deleted but the info in passwd wasn't updated, maybe this partition could not even exist.

I think this is the safest way to ensure we are setting the option to the right partition

[the remark this answers is not in the document]
Comment (marcusburghardt) [reply shared/templates/mount_option_home/bash.template:27]: I agree the proper partition should be ensured, but my concern is about the limitations of how this is ensured now, relying on the home dir existence. There are valid cases where the home dir folder may not present and the partition still needs to be remediated. Could be simply because the home dir was not created or was deleted. But could also be cases where `automount` is dynamically mounting the partition, etc.

The logic should be adjusted to not depend on the user home dir existence to configure the partition options. Maybe the approach of using the `df` could be changed. There is also a fragility with the `df` approach, as I mentioned in the other comment.

The logic could be something like this:
- Collect all home dirs from interactive users from /etc/passwd, as already done.
- Collect the mounted partitions.
- Check if any home dir is below any mounted partition.
- If so, apply the remediation in the partition, as already done.

If this logic is too complex and not feasible for the moment, at least include a warning in these rules should be included, informing these limitations when this new template is used.

File: linux_os/guide/system/permissions/partitions/mount_option_home_nosuid/tests/fstab.fail.sh (hunk 6 above)
[the remark this answers is not in the document]
Comment (marcusburghardt) [reply linux_os/guide/system/permissions/partitions/mount_option_home_nosuid/tests/fstab.fail.sh:3]: The case is that the templated script and its supposedly equivalent in the rule level behave differently. So, they are not equivalent. The existing tests should not fail. The existing test scenarios are failing because the remediation is not capable to fix the partition settings if there isn't a home dir created.

File: shared/templates/mount_option_home/bash.template (hunk 7 above)
[the remark this answers is not in the document]
Comment (Xeicker) [reply shared/templates/mount_option_home/bash.template:27]: There is an issue with your suggestion
If the right partition is below a wrong partition, there is the possibility to modify that wrong partition.
Taking the example in one of the first comments:
Let say we have a partition mounted in `/var` and there is another partition usually mounted in `/var/tmp` but not mounted now.
If there is an user with home directory in `/var/tmp/home_user1`, there is no way to ensure that the script will always find the partition usually mounted in `/var/tmp` and it then will modify the partition in `/var`

I'll set the warning for now

File: linux_os/guide/system/permissions/partitions/mount_option_home_nosuid/tests/fstab.fail.sh (hunk 6 above)
[the remark this answers is not in the document]
Comment (Xeicker) [reply linux_os/guide/system/permissions/partitions/mount_option_home_nosuid/tests/fstab.fail.sh:3]: Well is equivalent in the sense that the conf in fstab is what makes OVAL fail. This test is not correctly set for the template, precisely because it has a different behavior, if there are no users with home directory under `/home` the test won't result in fail, and that is expected.

Even if we want to fix that in remediation, I think we should set a specific test to validate more explicitly that scenario

**Review by marcusburghardt — APPROVED** [review]
There are many corner cases which this new template is limited to properly treat. On the other hand, they are not easy to be solved. Also, this PR had good improvements since its creation and the discussions are all registered. So, since a warning message was included in the respective rules to ensure the users aware of these limitations, I am approving the PR now. I also encourage more contributions and ideas to keep improving this template so it could be more widely used in the future. Thanks @Xeicker
